content/blog/next-level-iac-briding-the-declarative-gap/index.md

```diff
@@ -0,0 +1,164 @@
+- [**Deployment Engine**][pulumi-engine-docs]: A process that executes requests to create/modify infrastructure.
+- [**Resource Providers**][pulumi-providers-docs]: Plugins that interface with cloud services like AWS, Azure, and others.
+
+These components interact together to manage your infrastructure across multiple cloud providers using a consistent programming and execution model.
+
+{{< figure src="pulumi-architecture-simplified.png" caption="Figure: Pulumi's Program, Deployment Engine, and Resource Provider architecture">}}
+
+Let's get into an example of how information flows back forth between these components. Suppose you want to write a Pulumi program to create an AWS S3 Bucket to [host a static website][pulumi-static-website-example]. When that bucket is created, it will be given a unique indentifier that you need to use in other parts of your infrastructure, in order to reference the bucket. Every time you create the bucket the identifier will change, so we need to handle this as a variable not a fixed string.
+
+To do that in a Pulumi program, you can create a bucket and then refer to its `id` property later, like this:
+
+{{< chooser language "typescript,python" />}}
+
+{{% choosable language typescript %}}
+
+```typescript {hl_lines=[8]}
+import * as pulumi from '@pulumi/pulumi';
+import * as aws from '@pulumi/aws';
+
+// Create an S3 bucket
+const bucket = new aws.s3.Bucket("bucket");
+
+// Export the name of the bucket
+export const bucketName = bucket.id;
+```
+
+{{% /choosable %}}
+
+{{% choosable language python %}}
+
+```python {hl_lines=[8]}
+import pulumi
+import pulumi_aws as aws
+
+# Create an S3 bucket
+bucket = aws.s3.Bucket('bucket')
+
+# Export the name of the bucket
+pulumi.export('bucket_name', bucket.id)
+```
+
+{{% /choosable %}}
+
+When this program runs, the call to create the `bucket` object will issue a request to the *deployment engine*. The deployment engine will load the *resource provider* for AWS, and ask it to create the S3 bucket. The AWS resource provider then issues the request to the AWS API, which will create the bucket and then return the `id` of the newly created resource. The provider will return that to the deployment engine, which will then *set the realized value* on your Pulumi program's `bucket.id` property and then move on to creating anything that depends on that property.
+
+This all seems pretty straight-forward, except that Pulumi programs don't wait and imperatively do each operation, one at a time. Instead the deployment engine will run them in parallel, only waiting in cases where the output of one operation is needed to run the next operation. This can create complex structures with some operations needing to wait for inputs from other operations, and some that don't, all finishing at different times, depending on how long the cloud service takes to process that request.
+
+## Inputs and Outputs as asynchronous futures
```

Comment:
I think Inputs and Outputs are Pulumi terms. So perhaps you should make the title more generic to the use case being expressed. So maybe its really "Variables with asynchronous futures" or something?

Reply:
Well, inputs/outputs are pretty common/general terms but the way we use them in Pulumi is specific. That's sort of what I'm getting into in this section... explaining that they are not just simple input/output values/variables, but actually futures that model both asynchronous value access as well as relationship dependency between components within Pulumi, and that they too function interact in a specific and magical way that belies the simplicity of the terminology and API of Pulumi's Inputs/Outputs.